swap/htlc.go

```diff
@@ -190,49 +280,213 @@ func swapHTLCScript(cltvExpiry int32, senderHtlcKey,
+
+	builder.AddOp(txscript.OP_SIZE)
+	builder.AddInt64(0x20)
+	builder.AddOp(txscript.OP_EQUALVERIFY)
+	builder.AddOp(txscript.OP_HASH160)
+	builder.AddData(input.Ripemd160H(swapHash[:]))
+	builder.AddOp(txscript.OP_EQUALVERIFY)
+	builder.AddOp(txscript.OP_1)
+
+	builder.AddOp(txscript.OP_CHECKSEQUENCEVERIFY)
+
+	builder.AddOp(txscript.OP_ENDIF)
+
+	return builder.Script()
+}
+
+// newHtlcScript instantiates the htlcScript based on the passed version.
+// Returns ErrInvalidScriptVersion if the passed script version is unknown.
+func newHtlcScript(scriptVersion ScriptVersion, cltvExpiry int32, senderKey,
+	receiverKey [33]byte, hash lntypes.Hash) (HtlcScript, error) {
+
+	switch scriptVersion {
+	case HtlcV1:
+		script, err := swapHTLCScriptV1(
+			cltvExpiry, senderKey, receiverKey, hash,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		return &HtlcScriptV1{
+			script: script,
+		}, nil
+
+	case HtlcV2:
+		script, err := swapHTLCScriptV2(
+			cltvExpiry, senderKey, receiverKey, hash,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		return &HtlcScriptV2{
+			script: script,
+		}, nil
 	}
 
+	return nil, ErrInvalidScriptVersion
+}
+
+// HtlcScriptV1 encapsulates the htlc v1 script.
+type HtlcScriptV1 struct {
+	script []byte
+}
+
+// genSuccessWitness returns the success script to spend this htlc with
+// the preimage.
+func (h *HtlcScriptV1) genSuccessWitness(receiverSig []byte,
+	preimage lntypes.Preimage) wire.TxWitness {
+
 	witnessStack := make(wire.TxWitness, 3)
 	witnessStack[0] = append(receiverSig, byte(txscript.SigHashAll))
 	witnessStack[1] = preimage[:]
-	witnessStack[2] = h.Script
+	witnessStack[2] = h.script
 
-	return witnessStack, nil
+	return witnessStack
 }
 
-// IsSuccessWitness checks whether the given stack is valid for redeeming the
+// genTimeoutWitness returns the timeout script to spend this htlc after
+// timeout.
+func (h *HtlcScriptV1) GenTimeoutWitness(senderKey [33]byte,
+	senderSig []byte) wire.TxWitness {
+
+	witnessStack := make(wire.TxWitness, 3)
+	witnessStack[0] = append(senderSig, byte(txscript.SigHashAll))
+	witnessStack[1] = []byte{0}
+	witnessStack[2] = h.script
+
+	return witnessStack
+}
+
+// isSuccessWitness checks whether the given stack is valid for redeeming the
 // htlc.
-func (h *Htlc) IsSuccessWitness(witness wire.TxWitness) bool {
+func (h *HtlcScriptV1) IsSuccessWitness(witness wire.TxWitness) bool {
 	if len(witness) != 3 {
 		return false
 	}
 
 	isTimeoutTx := bytes.Equal([]byte{0}, witness[1])
+	return !isTimeoutTx
+}
+
+// GetScript returns the htlc script.
+func (h *HtlcScriptV1) GetScript() []byte {
+	return h.script
+}
+
+// MaxSuccessWitnessSize returns the maximum success witness size.
+func (h *HtlcScriptV1) MaxSuccessWitnessSize() int {
+	// Calculate maximum success witness size
+	//
+	// - number_of_witness_elements: 1 byte
+	// - receiver_sig_length: 1 byte
+	// - receiver_sig: 73 bytes
+	// - preimage_length: 1 byte
+	// - preimage: 33 bytes
```

Comment:
preimage is 32 bytes? (pre-existing I think)

Reply:
also in v2

Reply:
nice catch!